### PR TITLE
Enhancements to the Strimzi Kafka Operator for managing shared volume mounts,

### PR DESCRIPTION
**Pull request by rao2100**
Proposal to introduce enhancements to the Strimzi Kafka Operator for managing shared volume mounts, improving task resiliency across nodes.

The changes for the proposal in the PR (https://github.com/strimzi/strimzi-kafka-operator/pull/9761)

### CODE REVIEW

File: 069-shared-volume-mount.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+1) **mountPath:** A string property to specify the container directory for mounting the storage volume. While the default is /var/lib/kafka/data, this allows for custom path specification. By default, it will try to mount read write many storage if accessMode is not specified.
+
+2) **accessMode:** An optional string property defining the storage access mode, determining permissions like read-only or read-write.
+
+These changes, tested successfully with Read-Write-Many on Ceph-block storage and NFS, provide greater control over volume configuration and usage, improving operator performance in distributed environments. 
+
+**Sample Test**
+
+ - This test case is to test the strimzi operator customization that allows mounting of shared volume across kafka-connect pods
```
Comment (scholzj) [inline 069-shared-volume-mount.md:35]: You should also explain how will this be cover in the various test environments -> how will you provide the shared storage for example?

File: 069-shared-volume-mount.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Compatibility
+
+The proposed enhancement to the Strimzi Kafka Operator has been designed with both future and backward compatibility in mind. For future compatibility, the feature is built to be flexible and adaptable to accommodate potential changes or additions to the storage and access modes. This ensures that the feature remains relevant and useful as the system evolves. For backward compatibility, the feature maintains the existing functionality and behavior of the operator when the new options are not specified. This means that existing projects can continue to operate without any changes, ensuring a smooth transition and minimizing disruption to current operations..
+
+## Rejected alternatives
+
+While formulating this proposal, several alternatives were considered but ultimately rejected for various reasons:
+
+- **Using Local Storage**: Initially, the idea of using local storage for each pod was considered. However, this was rejected because it would not allow data sharing across pods, which is a crucial requirement for ensuring high availability and fault tolerance in distributed systems.
```
Comment (scholzj) [inline 069-shared-volume-mount.md:104]: I think it is fine to _reject_ local storage. But I do not think shared persistent volumes are a good basis for distributed systems as you suggest as they cause many limitations as to how _distributed_ they can be.

File: 069-shared-volume-mount.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+- **Static Volume Mounts**: Another alternative was to use static volume mounts with predefined paths. This was rejected due to its lack of flexibility. It would not allow custom specification of the container directory for storage volume mounting, which is necessary for accommodating different system configurations and use cases.
```
Comment (scholzj) [inline 069-shared-volume-mount.md:105]: I don't think it is really clear what `Static Volume Mounts` are.

File: 069-shared-volume-mount.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
+# Shared Volume Mount Changes
+
+The proposed update to the Strimzi Kafka Operator introduces management of shared volume mounts for kafka connect cluster,
```
Comment (scholzj) [inline 069-shared-volume-mount.md:3]: We need to have a reusable solution that covers all the different use cases for mounting some volumes. So you should not focus just on your immediate need. We have to make sure this change can survive the _next proposal_ from someone else with different requirement.

**Review by scholzj — COMMENTED** [review]
Thanks for the proposal. As I already mentioned in the PR, we have to make sure it matches all the various usecases mentioned in the existing Strimzi issues such as https://github.com/strimzi/strimzi-kafka-operator/issues/3693. And not just your single use-case. So you should cover all of them in this proposal and explain how they will be fulfilled.

My (personal) view is that we do not want to extend the existing `.storage` APi which is designed in sync withj Kafka's requirement. Instead, I think the right way might be to extend the Pod and Container template to allow adding volumes and mount points?

File: 069-shared-volume-mount.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+Support for Pod Running Logs: Ability to dump kafka connect logs to storage volume instead of just stdout.
+GC Logs Management: Implement a JVM option to output GC logs to a file instead of stdout when GC logging is enabled. This will help in identifying JVM hangs and keep stdout uncluttered.
+Additional diagnostics: Explore the possibility of mounting additional volumes into the Kafka Connect pods beyond ConfigMaps or Secrets. This could facilitate writing event data to parquet files for later analysis.
+
+## Proposal
+
+This change introduces enhancements to the Strimzi Kafka Operator for managing shared volume mounts, improving task resiliency across nodes. Key changes include:
+
+1) **mountPath:** A string property to specify the container directory for mounting the storage volume. While the default is /var/lib/kafka/data, this allows for custom path specification. By default, it will try to mount read write many storage if accessMode is not specified.
```
Comment (ppatierno) [inline 069-shared-volume-mount.md:27]: It's not just about where the volume is mounted but this path is also related to the Kafka broker configuration where putting the logs via `log.dirs` param. 
Also how it's going to work in terms of backward compatibility if for an already running cluster you change this path. What happens to the log segments?

File: 069-shared-volume-mount.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+2) **accessMode:** An optional string property defining the storage access mode, determining permissions like read-only or read-write.
```
Comment (ppatierno) [inline 069-shared-volume-mount.md:29]: can you elaborate what's the rationale behind this need?

File: 069-shared-volume-mount.md (hunk 4 above)
Comment (ppatierno) [reply 069-shared-volume-mount.md:3]: I am with Jakub on this.

**Review by ppatierno — COMMENTED** [review]
I left a few comments but one thing you should do is splitting all the sentences, one per line. It's really hard to make comments on a specific sentence when there are more and how GitHub works.

**Comment by scholzj** [comment]
@rao2100 Do you plan to incorporate the comments into the proposal? Or should this be closed?

**Comment by scholzj** [comment]
Discussed on the community call on 2.5.2024: Should be closed as there does not seem to be any progress. Feel free to reopen it / open new PR with the comments incorporated.

**Pull request closed without merging**